sdk/rust/vaa_payload/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+//! Serialize and deserialize rust values from the VAA payload wire format.
+//!
+//! As of this writing (June, 2022) there is no proper specification for the VAA payload wire
+//! format so this implementation has mostly been reverse engineered from the existing messages.
+//! While the rest of this document talks about how various types are represented on the wire this
+//! should be seen as an explanation of how things are implemented *in this crate* and not as
+//! official documentation. In cases where the serialization of a payload produced by this crate
+//! differs from the one use by the wormhole contracts, the serialization used by the actual
+//! contract is considered the canonical serialization.
+//!
+//! Unless you want to interact with existing wormhole VAA payloads, this crate is probably not what
+//! you are looking for. If you are simply using the wormhole bridge to send your own payloads then
+//! using a schema with auto-generated code (like protobufs or flatbuffers) is probably a better
+//! choice.
+//!
+//! ## Wire format
+//!
+//! The VAA payload wire format is not a self-describing format (unlike json and toml). Therefore it
+//! is necessary to know the type that needs to be produced before deserializing a byte stream.
+//!
+//! The wire format currently supports the following primitive types:
+//!
+//! ### `bool`
+//!
+//! Encoded as a single byte where a value of 0 indicates false and 1 indicates true. All other
+//! values are invalid.
+//!
+//! ### Integers
+//!
+//! `i8`, `i16`, `i32`, `i64`, `i128`, `u8`, `u16`, `u32`, `u64`, and `u128` are all supported and
+//! encoded as full-width big-endian integers, i.e., `i16` is 2 bytes, `u64` is 8 bytes, etc.
+//!
+//! ### `char`
+//!
+//! Encoded as a big-endian `u32`, with the additional restriction that it must be a valid [`Unicode
+//! Scalar Value`](https://www.unicode.org/glossary/#unicode_scalar_value).
+//!
+//! ### Sequences
+//!
+//! Variable length heterogeneous sequences are encoded as a single byte length followed by the
+//! concatenation of the serialized form of each element in the sequence. Note that this means that
+//! sequences cannot have more than 255 elements. Additionally, during serialization the length must
+//! be known ahead of time.
+//!
+//! ### Byte arrays - `&[u8]`, `Vec<u8>`, and `Cow<'a, [u8]>`
+//!
+//! Byte arrays are treated as a subset of variable-length sequences and are encoded as a single
+//! byte length followed by that many bytes of data. Again, since the length of the byte array has
+//! to fit in a single byte it cannot be longer than 255 bytes.
```

Comment:
fields like the VAA payload don't have a length field and can have unlimited length (till the end of the packet)

Reply:
That sounds good